Develop

Vintros opened the pull request.

Vintros committed:
fix: добавить каскадное удаление

BucketOnHead committed:
fix: HashSet<Genre> -> TreeSet<Genre>

BucketOnHead committed:
add: новая функциональность - удаление

BucketOnHead committed:
add: новая функциональность - удаление

BucketOnHead committed:
add: ./db/ в .gitignore

BucketOnHead committed:
update: экспериментальный скрипт

BucketOnHead committed:
update: экспериментальный скрипт

BucketOnHead committed:
update: экспериментальный скрипт

Vintros committed:
"feat: добавить функциональность отзывов"

Vintros committed:
Merge branch 'develop' into add-reviews

# Conflicts:
#	src/main/resources/schema.sql

BucketOnHead committed:
Merge pull request #6 from Vintros/add-remove-endpoint

Add remove endpoint

Vintros committed:
Merge branch 'develop' into add-reviews

BucketOnHead committed:
add: модель Director

BucketOnHead committed:
add: шаблон DirectorController

BucketOnHead committed:
add: шаблон DirectorService

BucketOnHead committed:
add: шаблон хранилища режиссёров

Vintros committed:
"refactor: в Review заменить мапу лайков на значение useful, получать из БД Review одним запросом"

Vintros committed:
"style: Актуализировать ER-диаграмму"

Vintros committed:
"style: Актуализировать ER-диаграмму"

BucketOnHead committed:
add: проверка name на пустоту

Vintros committed:
Merge pull request #9 from Vintros/add-reviews

Add reviews

BucketOnHead committed:
add: реализовать хранилище режиссёров

BucketOnHead committed:
add: реализовать сервис для режиссёров

BucketOnHead committed:
add: реализовать контроллер для режиссёров

BucketOnHead committed:
add: реализовать эндпоинты для режиссёров

BucketOnHead committed:
Merge branch 'develop' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/exception/ErrorHandler.java
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/validator/Validator.java
#	src/main/resources/schema.sql

BucketOnHead committed:
delete: тест на дату релиза в будущем

BucketOnHead committed:
style: подправить ляпы

BucketOnHead committed:
style: подправить ляпы

BucketOnHead committed:
delete: удалить дубликат таблицы

BucketOnHead committed:
update: использовать поиск через keyHolder вместо ручного

BucketOnHead committed:
add: идентификатор связи

marialyahovskaya committed:
Добавлен метод getCommonFilms в DbFilmStorage

marialyahovskaya committed:
Добавлен эндпоинт /films/common

ArtemRabus committed:
Первая итерация

ArtemRabus committed:
Первая итерация

BucketOnHead committed:
Merge pull request #10 from Vintros/add-director

Add director

ArtemRabus committed:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

AleksandrTimofeev1987 committed:
Функциональность Рекомендации.
feature: добавить endpoint в UserController
feature: добавить метод UserService.getRecommendations
feature: добавить метод DbUserStorage.getRecommendations
feature: добавить тест нового endpoint в UserControllerTest
refactor: добавить аннотацию @AutoConfigureTestDatabase в тестовые классы

AleksandrTimofeev1987 committed:
Функциональность Рекомендации.
debug: исправить ошибку при отсутствии лайков
debug: исправить ошибку при отсутствии пересечений лаков с дригими пользователями
debug: исправить ошибку при одиноковых лайках

AleksandrTimofeev1987 committed:
Функциональность Рекомендации.
refactor: перенести метод getRecommendations и дополнительные private методы в FilmStorage

ArtemRabus committed:
Вторая итерация

ArtemRabus committed:
Третья итерация (удалил аннотацию, спецом её не добавлял, видимо коряво смержил ветку develop)

ArtemRabus committed:
Третья итерация + (исправил опечатку)

AleksandrTimofeev1987 committed:
Merge pull request #12 from Vintros/add-recommendations

Добавлена функциональность Рекомендации

ArtemRabus committed:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

ArtemRabus committed:
Merge pull request #11 from Vintros/add-most-populars

Add most populars

AleksandrTimofeev1987 committed:
Функциональность Поиск.
feature: добавить end-point GET /fimls/search
feature: добавить метод FilmService.searchFilmsByTitleOrDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitle
feature: добавить метод DbFilmStorage.searchFilmsByDirector
feature: добавить метод DbFilmStorage.searchFilmsByTitleAndDirector

marialyahovskaya committed:
Merge branch 'develop' into add-common-films

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

marialyahovskaya committed:
Разрешен merge conflict

marialyahovskaya committed:
fix по замечаниям code review

marialyahovskaya committed:
Merge pull request #15 from Vintros/add-common-films

Добавлен функционал common films

AleksandrTimofeev1987 committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/FilmStorage.java

AleksandrTimofeev1987 committed:
Функциональность Поиск.
refactor: перенести валидацию поискового запроса в класс Validator
refactor: перенести инициализацию валидации поискового запроса в FilmService.searchFilmsByTitleOrDirector

Vintros committed:
"feat: Добавить функциональность ленты событий"

Vintros committed:
"fix: Перенести конструктор в аннотацию, исправить схему и диаграмму"

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после рувью.
refactor: переписать методы поиска в DbFilmStorage, чтобы избежать множественных обращений к БД

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после ревью.
style: переименовать параметр by на searchBy у метода FilmService.searchFilmsByTitleOrDirector
fix: обратить сортировку в методе DbFilmStorage.populateFilmsWithLikes

AleksandrTimofeev1987 committed:
Функциональность Поиск. Корректировка после ревью.
style: исправить опечатку в логировании метода FilmService.searchFilmsByTitleOrDirector

AleksandrTimofeev1987 committed:
Merge pull request #14 from Vintros/add-search

Функциональность Поиск.

Vintros committed:
Merge branch 'develop' into add-feed

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java

Vintros committed:
Merge pull request #17 from Vintros/add-feed

"feat: Добавить функциональность ленты событий"

Vintros committed:
"style: убраны неиспользуемые аннотации, логирование приведено к одному стилю, большие конструкторы заменены на аннотации, удален неиспользуемый метод extractUsersIdLiked"

Vintros committed:
"fix: вернул пару аннотаций"

BucketOnHead committed:
add: аннотация, проверяющая дату фильма

BucketOnHead committed:
delete: старая проверка на дату фильма

BucketOnHead committed:
Merge pull request #19 from Vintros/hotfix

Hotfix

Vintros committed:
"fix: исправлена проверка даты релиза"

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/validator/Validator.java:
```diff
@@ -68,4 +69,41 @@ public static void validateMpaId(Long id) {
+
+    public static void validateReview(Long id) {
+        try {
+            reviewStorage.getReviewById(id);
```
Разве это валидация данных? Это ведь просто проверка наличия данных в базе. Лучше реализовать ее сразу в репозитории, либо в сервисе.

Vintros replied:
Проверка наличия данных перенесена в хранилище

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/validator/Validator.java:
```diff
@@ -68,4 +69,41 @@ public static void validateMpaId(Long id) {
+
+    public static void validateDirector(Long id) {
+        try {
+            directorStorage.getDirectorById(id);
```
аналогично

Vintros replied:
Проверка наличия данных перенесена в хранилище

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/validator/Validator.java:
```diff
@@ -68,4 +69,41 @@ public static void validateMpaId(Long id) {
+            directorStorage.getDirectorById(director.getId());
+            throw new ExistsException("Режиссёр уже существует");
+        } catch (EmptyResultDataAccessException ignored) {
+        }
```
похоже на временную заплатку) Тоже лучше перенести в сервис такую проверку (а лучше в хранилище). Если режиссер уже существует, то надо кинуть исключение, нельзя такое игнорировать на мой взгляд

Vintros replied:
Изменен метод проверки отсутствия данных в БД, перенесен в хранилище )

Vintros committed:
"style: в ридми указана добавленная функциональность"

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/model/Review.java:
```diff
@@ -0,0 +1,21 @@
+public class Review {
+
+    private Long reviewId;
+    @NonNull
```
я бы поставила пустые строки между полями для четкого разделения (просто сейчас у вас получается, что аннотации идут вперемешку с полями). Это не обязательно, но код станет выглядеть лучше)

Vintros replied:
Сделано

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorStorage.java:
```diff
@@ -0,0 +1,55 @@
+     * @return Коллекция из всех режиссёров
+     * в хранилище
+     */
+    Collection<Director> getDirectors();
```
мне кажется, нет необходимости возвращать аж коллекцию, достаточно вернуть лист, например)

Vintros replied:
Сделано

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/director/DirectorStorage.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    void removeDirectorById(Long id);
+
+    Map<Long, List<Director>> getDirectorsByFilmsId();
```
оставшимся методам тоже неплохо было бы добавить описание

Vintros replied:
Сделано

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/feed/FeedStorage.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface FeedStorage {
+
+    void saveUserEvent(Event event);
```
давайте во все интерфейсы добавим описания?

Vintros replied:
Нет причины отказаться )

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -1,28 +1,31 @@
 public class DbFilmStorage implements FilmStorage {
 
     private final JdbcTemplate jdbcTemplate;
     private final GenreStorage genreStorage;
```
Хранилища не должны зависеть от хранилищ. Вся бизнес логика должна быть в сервисе, а какие-то специфичные вещи можно вытащить через запросы)

Vintros replied:
Логика вынесена в сервис, вызов хранилищ из хранилищ исключен

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -113,6 +149,41 @@ private void addGenres(Set<Genre> genres, Long id) {
         });
     }
 
+    private void addDirectors(Set<Director> directors, Long id) {
```
все приватные методы лучше перенести в конец класса, обычно они там располагаются)

Vintros replied:
Исправлено

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -113,6 +149,41 @@ private void addGenres(Set<Genre> genres, Long id) {
 
+    private void addDirectors(Set<Director> directors, Long id) {
+        String sqlQuery = "insert into directors (director_id, film_id) values (?, ?)";
+        for (Director director : directors) {
```
давайте попробуем вместо 2-10-100 запросов в бд (в цикле) сделать один запрос обновления сразу всех данных? Подсказка: jdbcTemplate.batchUpdate и new BatchPreparedStatementSetter() внутри)) Это будет более правильный подход)

Vintros replied:
Запросы объединены в один

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -113,6 +149,41 @@ private void addGenres(Set<Genre> genres, Long id) {
+
+    @Override
+    public List<Film> getRecommendations(Long id) {
+        final String sqlQuery = "SELECT USER_ID, FILM_ID FROM LIKES GROUP BY USER_ID, FILM_ID";
```
сразу видно, что код писали разные люди)) выше запросы строчными буквами, здесь заглавными) 
Это не замечание, но бросается в глаза разный стиль

Vintros replied:
Привели все запросы к одному виду )

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -127,15 +198,217 @@ private Film mapRowToFilm(ResultSet rs, int rowNum) throws SQLException {
 
-    private Map<Long, List<Long>> extractUsersIdLiked(ResultSet rs) throws SQLException {
-        Map<Long, List<Long>> usersIdLiked = new LinkedHashMap<>();
+    private Map.Entry<Long, Long> mapRowToMapEntryLongLong(ResultSet rs, int rowNum) throws SQLException {
```
лонглонг аж ;)

Vintros replied:
Отрезали лонги )

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java:
```diff
@@ -127,15 +198,217 @@ private Film mapRowToFilm(ResultSet rs, int rowNum) throws SQLException {
+    }
+
+    private static Long getMostIntersectionsUserId(Long id, Map<Long, ArrayList<Long>> data) {
+        Map<Long, Integer> frequency = new HashMap<>();
```
Наоврекодили. Цикл в цикле и куча ветвлений, я даже разобраться не смогла) Думаю, можно было реализовать проще.. Есть идеи? Либо расскажите, что тут хотели сделать))

Vintros replied:
Постарались упростить :)

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -48,53 +54,114 @@ public void removeLikeFilm(Long id, Long userId) {
+        log.info("Получен поисковый запрос: {}. Параметр поиска: {}", query, searchBy);
+        validateSearchParameter(searchBy);
+        List<Film> films;
+                switch (searchBy) {
```
требуется Ctrl+Alt+l

Vintros replied:
сделано

utausheva reviewed (commented):
Привет всем!
Итак. Работа была проделана огромная! Вы выбрали не самые простые задания, но сумели реализовать правильно гигантские запросы и сложную логику. Вы неплохо сработались, постарались подстроить свой стиль под командный. Отдельно отмечу отличную работу тимлида команды, очень оперативно добавлялись данные по запросу в слаке ;) 
Теперь к итогам первой итерации ревью:
Модель получилась без нареканий, спасибо за перечень реализованных фич и новую схему бд! Необходимо подправить момент с использованием хранилища в хранилище (желательно выносить бизнес логику в сервисы, а в дао просто доставать нужные данные и делать предобработку). Немного запутали хранилище фильмов, там получилось слишком много сложной логики, которая обычно делается в сервисе, настолько сложной, что другому разработчику придется потратить много времени на разбор такого кода... Можете устроить брейн шторм всей командой и отрефакторить, времени у вас много))) Это если будет желание) Обычно в хранилище реализуются простые CRUD операции) Положительно выделю еще базу данных: очень круто получилось использовать ON DELETE CASCADE вместо доп действий по удалению записей из таблиц ;)
Критичных замечаний немного, в основном пожелания)
Думаю, вы справитесь быстро, жду на следующей итерации)))

BucketOnHead committed:
update: Collection<Director> -> List<Director>

BucketOnHead committed:
update: логику заполнения таблицы directors

AleksandrTimofeev1987 committed:
style: отформатировать все SQL запросы
style: отформатировать метод FilmService.searchFilmsByTitleOrDirector
style: переименовать метод DbFilmStorage.mapRowToMapEntryLongLong в DbFilmStorage.mapRowToMapEntry
style: убрать redundant локальные переменные

Vintros committed:
"refactor: проверка наличия сущностей перенесена из валидатора в стореджи с вызовом из сервисов"

Vintros committed:
"style: отформатировал вид новых sql запросов"

Vintros committed:
Merge pull request #25 from Vintros/fix-something

"style: отформатировал вид новых sql запросов"

BucketOnHead committed:
Merge pull request #21 from Vintros/fix-directors

Fix directors

BucketOnHead committed:
add: комментарии для интерфейса DirectorStorage

BucketOnHead committed:
add: комментарии для интерфейса FeedStorage

BucketOnHead committed:
add: комментарии для интерфейса FilmStorage

BucketOnHead committed:
add: комментарии для интерфейса GenreStorage

BucketOnHead committed:
add: комментарии для интерфейса MpaStorage

BucketOnHead committed:
add: комментарии для интерфейса ReviewStorage

BucketOnHead committed:
add: комментарии для интерфейса UserStorage

BucketOnHead committed:
update: вынес дату рождения кино в локальную константу

BucketOnHead committed:
Merge branch 'develop' into add-interface-description

BucketOnHead committed:
update: подтянуть изменения + Collection<?> -> List<?>

BucketOnHead committed:
update: убрать ')'

AleksandrTimofeev1987 committed:
Merge branch 'develop' into fix_sql

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/director/DbDirectorStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java

AleksandrTimofeev1987 committed:
Merge with develop

AleksandrTimofeev1987 committed:
Merge pull request #22 from Vintros/fix_sql

Форматирование SQL и малые стилевые исправления

BucketOnHead committed:
style: перенести приватные методы вниз

BucketOnHead committed:
style: перенести приватные методы вниз

BucketOnHead committed:
style: перенести приватные методы вниз

BucketOnHead committed:
style: перенести приватные методы вниз

BucketOnHead committed:
style: перенести приватные методы вниз

AleksandrTimofeev1987 committed:
Упрощение метода getMostIntersectionsUserId:
fix: упростить алгоритм поиска количества совпадений лайков в DbFilmStorage.getMostIntersectionsUserId
refactor: усложнить тест endpoint'а /{id}/recommendations UserControllerTest.shouldReturn200andListOnGetRecommendationsWhenValidRequest

AleksandrTimofeev1987 committed:
Упрощение метода getMostIntersectionsUserId:
style: упростить проверку mostIntersectionsUser

BucketOnHead committed:
Merge pull request #26 from Vintros/add-interface-description

Add interface description

BucketOnHead committed:
Merge pull request #27 from Vintros/fix-move-private-methods

Перенести приватные методы в конец класса

Vintros committed:
Merge branch 'develop' into fix-validator

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/review/DbReviewStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/user/UserStorage.java

Vintros committed:
"style: исправить форматирование и названия методов по замечаниям, добавить описание методов"

AleksandrTimofeev1987 committed:
Merge branch 'develop' into fix_intersections_search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/DbFilmStorage.java

AleksandrTimofeev1987 committed:
Упрощение метода getMostIntersectionsUserId

AleksandrTimofeev1987 committed:
Merge pull request #28 from Vintros/fix_intersections_search

Упрощение метода getMostIntersectionsUserId

Vintros committed:
"refactor: извемена проверка отсутствия сущности для передачи на проверку только id.
style:конкретизированы названия методов проверки"

Vintros committed:
Merge pull request #24 from Vintros/fix-validator

"refactor: проверка наличия сущностей перенесена из валидатора в стор…

ArtemRabus committed:
All logs and exceptions are translated into English

ArtemRabus committed:
Merge branch 'develop' into logs-rus-to-eng

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java
#	src/main/java/ru/yandex/practicum/filmorate/validator/Validator.java

ArtemRabus committed:
All logs and exceptions are translated into English

ArtemRabus committed:
Some fixes

ArtemRabus committed:
Merge pull request #29 from Vintros/logs-rus-to-eng

All logs and exceptions are translated into English

Vintros committed:
"refactor: перенести логику из стореджа в сервис"

Vintros committed:
"fix: merge изменений из develop"

Vintros committed:
"style: удалить устаревший комментарий"

Vintros committed:
Merge pull request #30 from Vintros/fix-storage

Fix storage

utausheva reviewed (approved):
😍😍😍
Очень круто получилось!
Еще раз скажу, что вы очень крутая команда, справились с большим количеством замечаний! Удачи вам всем на следующих спринтах! 
Можно мерджить))
